planning.py

```diff
@@ -1435,3 +1587,105 @@ def go_to_sfo():
+    """
+    
+    def __init__(self, action, precond , effect, duration =0, consume = None, use = None):
+        super().__init__(action, precond, effect, duration, consume, use)
+
+
+    def convert(self, clauses):
+        """Converts strings into Exprs""" 
+        lib = {'~': 'Not', 
+                '$+': 'PosYes',
+               '$-': 'PosNot',
+               '$$' : 'PosYesNot'}
+
+        if isinstance(clauses, Expr):
+            clauses = conjuncts(clauses)
+            for i in range(len(clauses)):
+                for ch in lib.keys():
+                    if clauses[i].op == ch:
+                        clauses[i] = expr( lib[ch] + str(clauses[i].args[0]))
+
+        elif isinstance(clauses, str):
+            for ch in lib.keys():
+                clauses = clauses.replace(ch, lib[ch])
+            if len(clauses) > 0:
+                clauses = expr(clauses)
+
+            try:
+                clauses = conjuncts(clauses)
+            except AttributeError:
+                pass
+
+        return clauses
+
+        
+        
+
+    def angelic_action(self):
+        """
+        example: 
+        """
+        #print(self.effect)
+
+        effects=[[]]
+        for clause in self.effect:
+            if clause.op[:9] == 'PosYesNot':
+                effects = effects*3
+                n=3
+                w=9
+            elif clause.op[:6] == 'PosYes':
+                effects =effects* 2
+                n=2
+                w=6
+            elif clause.op[:6] == 'PosNot':
+                effects = effects* 2
+                n=2
+                w=3
+            else:
+                n=1
+                w=0
+            it=range(1)
+            if len(effects)!=0:
+                it = range(len(effects)//n)
+            for i in it:
+                if effects[i]:
+                    if clause.args: 
+                        effects[i] = expr(str(effects[i]) + '&' + str(Expr(clause.op[w:],clause.args[0])))
+                        if n==3:
+                            effects[i+len(effects)//3]= expr(str(effects[i+len(effects)//3]) + '&' + str(Expr(clause.op[6:],clause.args[0])))
+                    else: 
+                        effects[i] = expr(str(effects[i]) + '&' + str(expr(clause.op[w:])))
+                        if n==3: effects[i+len(effects)//3] = expr(str(effects[i+len(effects)//3]) + '&' + str(expr(clause.op[6:])))
+
+                else: 
+                    if clause.args: 
+                        effects[i] = Expr(clause.op[w:], clause.args[0])
+                        if n==3: effects[i+len(effects)//3] = Expr(clause.op[6:], clause.args[0])
+
+                    else: 
+                        effects[i] = expr(clause.op[w:])
+                        if n==3: effects[i+len(effects)//3] = expr(clause.op[6:])
+            #print('effects',  effects)
+        
+        return [ HLA(Expr(self.name, self.args), self.precond, effects[i] ) for i in range(len(effects)) ]
+
```

Comment:
Some comments in the function angelic_action would be really nice. It might be nice to have a separate function that computes and returns n & w. Again this would allow you to test that aspect of the code and improve readability of angelic_action.

Reply:
I have added comments explaining the functionality of angelic_action, 
but maybe I need to add some more